clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -256,11 +256,13 @@ public void poll(long timeout, long now, PollCondition pollCondition) {
      * nor will it execute any delayed tasks.
      */
     public void pollNoWakeup() {
-        disableWakeups();
-        try {
-            poll(0, time.milliseconds(), null);
-        } finally {
-            enableWakeups();
+        synchronized (this) {
```

Comment:
I think I prefer the approach using the extra parameter in `poll()` if we can convince ourselves that it is safe (i.e. that there are no completion handlers which could either miss an expected wakeup or propagate one where it wasn't expected). I looked over the code and couldn't find any such cases.